syncer/syncer.go

```diff
@@ -2319,7 +2319,46 @@ func (s *Syncer) handleRowsEvent(ev *replication.RowsEvent, ec eventContext) err
 	return nil
 }
 
-func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, originSQL string) error {
+type queryEventContext struct {
+	*eventContext
+
+	p         *parser.Parser // used parser
+	ddlSchema string         // used schema
+	originSQL string         // before split
+	// split multi-schema change DDL into multiple one schema change DDL due to TiDB's limitation
+	splitedDDLs    []string // after split before online ddl
+	appliedDDLs    []string // after onlineDDL apply if onlineDDL != nil and track, before route
+	needHandleDDLs []string // after route
+
+	ddlInfo         *shardingDDLInfo
+	needTrackDDLs   []trackedDDL
+	sourceTbls      map[string]map[string]struct{} // db name -> tb name
+	onlineDDLTables map[string]*filter.Table
+}
+
+func (qec *queryEventContext) String() string {
+	var startLocation, currentLocation, lastLocation string
+	if qec.startLocation != nil {
+		startLocation = qec.startLocation.String()
+	}
+	if qec.currentLocation != nil {
+		currentLocation = qec.currentLocation.String()
+	}
+	if qec.lastLocation != nil {
+		lastLocation = qec.lastLocation.String()
+	}
+	var needHandleDDLs, shardingReSync string
+	if qec.needHandleDDLs != nil {
+		needHandleDDLs = strings.Join(qec.needHandleDDLs, ",")
+	}
+	if qec.shardingReSync != nil {
+		shardingReSync = qec.shardingReSync.String()
+	}
+	return fmt.Sprintf("{schema: %s, originSQL: %s, startLocation: %s, currentLocation: %s, lastLocation: %s, re-sync: %s, needHandleDDLs: %s}",
```

Comment:
If we want to splice a large number of strings, we can use `string buffer` for better performance
ref  https://stackoverflow.com/questions/1760757/how-to-efficiently-concatenate-strings-in-go

Reply:
maybe `strings.Join` is ok. https://segmentfault.com/a/1190000040275250

Reply:
Why don't we use "+" or builder to do this?

Reply:
> Why don't we use "+" or builder to do this?

The implementation of `join` is the `builder`